pkg/controller/networkpolicy/controller_querier_replier.go

```
@@ -0,0 +1,182 @@
+}
+
+// EndpointQueryResponse is the reply struct for QueryNetworkPolicies
+type EndpointQueryResponse struct {
+	Endpoints []Endpoint `json:"endpoints,omitempty"`
+}
+
+// Endpoint holds response information for an endpoint following a query
+type Endpoint struct {
+	Namespace string   `json:"namespace,omitempty"`
+	Name      string   `json:"name,omitempty"`
+	Policies  []Policy `json:"policies,omitempty"`
+	Rules     []Rule   `json:"rules,omitempty"`
+}
+
+type PolicyRef struct {
+	Namespace string    `json:"namespace,omitempty"`
+	Name      string    `json:"name,omitempty"`
+	UID       types.UID `json:"uid,omitempty"`
+}
+
+// Policy holds network policy information to be relayed to client following query endpoint
+type Policy struct {
+	PolicyRef
+	selector metav1.LabelSelector `json:"selector,omitempty"`
+}
+
+type Rule struct {
+	PolicyRef
+	Direction networkingv1beta1.Direction `json:"direction,omitempty"`
+	RuleIndex int                         `json:"ruleindex,omitempty"`
+}
+
+// NewNetworkPolicyController returns a new *NetworkPolicyController.
+func NewEndpointQueryReplier(networkPolicyController *NetworkPolicyController) *EndpointQueryReplier {
+	n := &EndpointQueryReplier{
+		networkPolicyController: networkPolicyController,
+	}
+	return n
+}
+
+//QueryNetworkPolicies returns kubernetes network policy references relevant to the selected network endpoint. Relevant
+//policies fall into three categories: applied policies (Policies in Endpoint type) are policies which directly apply to
+//an endpoint, egress and ingress rules (Rules in Endpoint type) are policies which reference the endpoint in an ingress/
+//egress rule respectively.
+func (eq EndpointQueryReplier) QueryNetworkPolicies(namespace string, podName string) (*EndpointQueryResponse, error) {
+	// check if namespace and podName select an existing pod
+	_, err := eq.networkPolicyController.podInformer.Lister().Pods(namespace).Get(podName)
+	if err != nil {
+		return nil, nil
```

do you want to return the error in this case?

---

A nil error with a nil return value is handled as a specific case in the handler. It was a while ago but I believe we had a conversation about this @antoninbas. Should I handle the specific error rather than return nil?

---

I think the current code is fine. Maybe the function comment should explicitly call out that we return `nil` and no error if the Namespace and Pod name do not match a known Pod.